src/Parser.php

```diff
@@ -404,50 +378,94 @@ protected function parseOutline()
+                // end of this scenario and be about to start a new one.
+                if ($this->predictNextTaggedNodeType() !== 'Examples') {
+                    break;
+                }
+            }
+
             $node = $this->parseExpression();
 
-            if ($node instanceof StepNode) {
-                $steps[] = $this->normalizeStepNodeKeywordType($node, $steps);
+            if ($steps === [] && is_string($node)) {
+                // Free text is only allowed before the first step (or when parsing Examples: which is done elsewhere)
+                $text = preg_replace('/^\s{0,' . ($token['indent'] + 2) . '}|\s*$/', '', $node);
+                $title .= "\n" . $text;
                 continue;
             }
 
-            if ($node instanceof ExampleTableNode) {
-                $examples[] = $node;
-
+            if ($node === "\n") {
                 continue;
             }
 
-            if (count($steps) === 0 && is_string($node)) {
-                $text = preg_replace('/^\s{0,' . ($token['indent'] + 2) . '}|\s*$/', '', $node);
-                $title .= "\n" . $text;
+            if ($examples === [] && $node instanceof StepNode) {
+                // Steps are only allowed before the first Examples table (if any)
+                $steps[] = $this->normalizeStepNodeKeywordType($node, $steps);
                 continue;
             }
 
-            if ($node === "\n") {
+            if ($node instanceof ExampleTableNode) {
+                // NB: It is valid to have a Scenario with Examples: but no Steps
+                $examples[] = $node;
                 continue;
             }
 
-            throw new UnexpectedParserNodeException('Step or Examples table', $node, $this->file);
+            throw new UnexpectedParserNodeException(
+                match ($examples) {
+                    [] => 'Step, Examples table, or end of Scenario',
+                    default => 'Examples table or end of Scenario',
+                },
+                $node,
+                $this->file,
+            );
         }
 
-        if (count($examples) === 0) {
-            throw new ParserException(sprintf(
-                'Outline should have examples table, but got none for outline "%s" on line: %d%s',
-                rtrim($title),
-                $line,
-                $this->file ? ' in file: ' . $this->file : ''
-            ));
+        if ($examples !== []) {
+            return new OutlineNode(rtrim($title) ?: null, $tags, $steps, $examples, $keyword, $line);
         }
 
-        return new OutlineNode(rtrim($title) ?: null, $tags, $steps, $examples, $keyword, $line);
+        return new ScenarioNode(rtrim($title) ?: null, $tags, $steps, $keyword, $line);
+    }
+
+    /**
+     * Peek ahead to find the node that the current tags belong to.
+     *
+     * @throws UnexpectedTaggedNodeException if there is not a taggable node
+     */
+    private function predictNextTaggedNodeType(): string
```

Comment:
This function actually does two things: validates that the next tagged node is correct and returns it. But the function name does not reflect the validation part, so when we call it and don't use the return value it looks weird. Also `predict` sounds a bit magic in my head 😆 . What about `validateAndGetNextTaggedNodeType()`?